pkg/controlsvc/controlsvc_test.go

```diff
@@ -0,0 +1,647 @@
+				t.Errorf("should be nil")
+			}
+			if testCase.message != "" && err.Error() != "cannot write message" {
+				t.Errorf("%s %s", testCase.name, err)
+			}
+		})
+	}
+}
+
+func TestSockControlBridgeConn(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCon := mock_controlsvc.NewMockConn(ctrl)
+	mockUtil := mock_controlsvc.NewMockUtiler(ctrl)
+	mockCopier := mock_controlsvc.NewMockCopier(ctrl)
+	logger := logger.NewReceptorLogger("")
+
+	sockControl := controlsvc.NewSockControl(mockCon, mockUtil, mockCopier)
+
+	bridgeConnTestCases := []struct {
+		name          string
+		message       string
+		expectedCalls func()
+	}{
+		{
+			name:    "without message and no error",
+			message: "",
+			expectedCalls: func() {
+				mockUtil.EXPECT().BridgeConns(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any())
+			},
+		},
+		{
+			name:    "with message and error",
+			message: "message",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New("terminated"))
+			},
+		},
+	}
+
+	for _, testCase := range bridgeConnTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			testCase.expectedCalls()
+			err := sockControl.BridgeConn(testCase.message, mockCon, "test", logger)
+
+			if testCase.message == "" && err != nil {
+				t.Errorf("should be nil")
+			}
+			if testCase.message != "" && err.Error() != "terminated" {
+				t.Errorf("stuff %v", err)
+			}
+		})
+	}
+}
+
+func TestSockControlReadFromConn(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCon := mock_controlsvc.NewMockConn(ctrl)
+	mockUtil := mock_controlsvc.NewMockUtiler(ctrl)
+	mockCopier := mock_controlsvc.NewMockCopier(ctrl)
+
+	sockControl := controlsvc.NewSockControl(mockCon, mockUtil, mockCopier)
+
+	bridgeConnTestCases := []struct {
+		name          string
+		message       string
+		expectedCalls func()
+		expectedError bool
+		errorMessage  string
+	}{
+		{
+			name:    "without message and copier error",
+			message: "",
+			expectedCalls: func() {
+				mockCopier.EXPECT().Copy(gomock.Any(), gomock.Any()).Return(int64(0), errors.New("read from conn copy error"))
+			},
+			expectedError: true,
+			errorMessage:  "read from conn copy error",
+		},
+		{
+			name:    "with message and no error",
+			message: "message",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New("read from conn write error"))
+			},
+			expectedError: true,
+			errorMessage:  "read from conn write error",
+		},
+		{
+			name:    "without message and no copier error",
+			message: "",
+			expectedCalls: func() {
+				mockCopier.EXPECT().Copy(gomock.Any(), gomock.Any()).Return(int64(0), nil)
+			},
+			expectedError: false,
+			errorMessage:  "",
+		},
+	}
+
+	for _, testCase := range bridgeConnTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			testCase.expectedCalls()
+			err := sockControl.ReadFromConn(testCase.message, mockCon)
+
+			if testCase.expectedError {
```

Comment:
Im having trouble understanding everything in this `if` block. 

It looks like you only need 1 `if` statement: 
```
if err != nil && err.Error() != testCase.errorMessage {
					printExpectedError(t, err)
				}
```